pss/message/message_test.go

```diff
@@ -0,0 +1,51 @@
+package message_test
+
+import (
+	"encoding/hex"
+	"fmt"
+	"testing"
+
+	"github.com/epiclabs-io/ut"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/ethersphere/swarm/pss/message"
+)
+
+func TestMessage(tx *testing.T) {
+	t := ut.BeginTest(tx, false) // set to true to generate test results
+	defer t.FinishTest()
+
+	// generate some test messages deterministically
+	for i, topicString := range someTopics {
+		flags := message.Flags{
+			Raw:       i&0x1 == 0,
+			Symmetric: i&0x3 == 0,
+		}
+
+		msg := message.New(flags)
+		msg.To = ut.RandomArray(i, common.AddressLength)
+		msg.Expire = uint32(i)
+		msg.Topic = message.NewTopic([]byte(topicString))
+		msg.Payload = ut.RandomArray(i*9361, i*10)
+
+		// test digest function:
+		digest := msg.Digest()
+		t.EqualsKey(fmt.Sprintf("msg%d-digest", i), hex.EncodeToString(digest[:]))
+
+		// test stringer:
+		st := msg.String()
+		t.EqualsKey(fmt.Sprintf("msg%d-string", i), st)
+
+		// Test RLP encoding:
+		bytes, err := rlp.EncodeToBytes(&msg)
+		t.Ok(err)
+		t.EqualsKey(fmt.Sprintf("msg%d-rlp", i), hex.EncodeToString(bytes))
```

Comment:
How did you generate the results for this, please?

Reply:
At the begining of the test method, there is a flag you can set to `true` which will generate all test results if you run the test. When you do that, the test will be marked as failed, but it will have generated all values. You can then revert the flag back to `false` and the stored results are verified against.

You would only do that if after changing the `Message` structure for some reason, e.g., a protocol upgrade, and want the tests to be updated automatically. This avoids having strings all over the code and having to manually edit them when there is the slightest change (e.g., incorporating `Expires` to the digest, which will happen soon 😉).

Results tested with `EqualsKey()` are stored under `testdata/<Test Name>/results.json` under the corresponding key, in this case `msg0-rlp`, `msg1-rlp`... and so on. 

(`testdata` folders are special and skipped by Go tools)